cachito/workers/pkg_managers/yarn.py

```diff
@@ -55,3 +64,293 @@ def get_yarn_proxy_repo_username(request_id):
+    """
+    Get the main package and dependencies based on the lock file.
+
+    If the lockfile contains non-registry dependencies, the lock file will be modified to use ones
+    in Nexus. Non-registry dependencies will have the "version_in_nexus" key set.
+
+    :param (str | Path) package_json_path: the path to the package.json file
+    :param (str | Path) yarn_lock_path: the path to the lock file
+    :return: a dictionary that has the keys "deps" which is the list of dependencies,
+        "lock_file" which is the lock file if it was modified (as a dict!), "package" which is the
+        dictionary describing the main package, and "package.json" which is the package.json file if
+        it was modified.
+    :rtype: dict
+    """
+    with open(package_json_path) as f:
+        package_json = json.load(f)
+
+    yarn_lock = pyarn.lockfile.Lockfile.from_file(str(yarn_lock_path)).data
+
+    try:
+        package = {
+            "name": package_json["name"],
+            "version": package_json["version"],
+            "type": "yarn",
+        }
+    except KeyError:
+        raise CachitoError("The package.json file is missing required data (name, version)")
+
+    # TODO: use a different allowlist for yarn? rename allowlist variable?
+    file_deps_allowlist = set(
+        get_worker_config().cachito_npm_file_deps_allowlist.get(package["name"], [])
+    )
+
+    deps, nexus_replacements = _get_deps(yarn_lock, file_deps_allowlist)
+
+    if nexus_replacements:
+        package_json_replaced = _replace_deps_in_package_json(package_json, nexus_replacements)
+        yarn_lock_replaced = _replace_deps_in_yarn_lock(yarn_lock, nexus_replacements)
+    else:
+        package_json_replaced = None
+        yarn_lock_replaced = None
+
+    return {
+        "package": package,
+        "deps": deps,
+        "package.json": package_json_replaced,
+        "lock_file": yarn_lock_replaced,
+    }
+
+
+def _replace_deps_in_package_json(package_json, nexus_replacements):
+    """
+    Replace non-registry dependencies in package.json with their versions in Nexus.
+
+    :param dict package_json: parsed package.json data
+    :param dict nexus_replacements: modified subset of yarn.lock data, a dict in the format:
+        {<dependency identifier>: <dependency info>}
+    :return: copy of package.json data with replacements applied (or None if no replacements match)
+    """
+    expanded_replacements = {}
+
+    for deps_identifier, nexus_replacement in nexus_replacements.items():
+        # The dependency identifier is a line that may contain multiple comma-separated
+        # dependencies (different ways to specify the same dependency)
+        deps = map(str.strip, deps_identifier.split(","))
+        for d in deps:
+            expanded_replacements[d] = nexus_replacement
+
+    package_json_new = copy.deepcopy(package_json)
+    modified = False
+
+    # TODO: optionalDependencies also exist, ignored for npm, safe to ignore for yarn?
```

Comment:
I believe it would be ok to track this one somehow and address it in the future when/if needed.

@arewm, do you need optional dependencies in any of your projects?

Reply:
We do not currently have any yarn projects, but we might need to start building grafana in the near future (which uses yarn). I haven't looked at what their build setup is yet.

Reply:
@chmeliik, we can ignore these for now since we are ignoring them for npm as well. We may revisit this in the future.